rbf-chunker/src/main.rs

```diff
@@ -0,0 +1,227 @@
+use std::fs::File;
+use std::io::prelude::*;
+use std::io::{Error, ErrorKind};
+use std::path::{Path, PathBuf};
+
+#[derive(Debug, Copy, Clone)]
+enum SectionType {
+    Unkown,
+    Header,
+    Footer,
+    Packet { crc: [u8; 2] },
+    PacketCRC,
+}
+
+#[derive(Debug, Copy, Clone)]
+struct SectionInfo {
+    offset: usize,
+    size: usize,
+    stype: SectionType,
+}
+
+impl SectionInfo {
+    fn new_header(size: usize) -> Self {
+        SectionInfo {
+            offset: 0,
+            size,
+            stype: SectionType::Header,
+        }
+    }
+
+    fn new_unkown(offset: usize, size: usize) -> Self {
+        SectionInfo {
+            offset,
+            size,
+            stype: SectionType::Unkown,
+        }
+    }
+
+    fn new_footer(offset: usize, size: usize) -> Self {
+        SectionInfo {
+            offset,
+            size,
+            stype: SectionType::Footer,
+        }
+    }
+
+    fn new_packet(offset: usize, size: usize, crc: [u8; 2]) -> Self {
+        SectionInfo {
+            offset,
+            size,
+            stype: SectionType::Packet { crc },
+        }
+    }
+
+    fn new_packet_crc(offset: usize) -> Self {
+        SectionInfo {
+            offset,
+            size: 2,
+            stype: SectionType::PacketCRC,
+        }
+    }
+}
+
+const RBF_SIZE: usize = 7020944;
+const PACKET_MIN_SIZE: usize = 916;
+const PACKET_MAX_SIZE: usize = 916;
+const DISCARD_MAX_SIZE: usize = PACKET_MAX_SIZE;
+const HEADER_SIZE: usize = 0x84;
+const FOOTER_SIZE: usize = 0x268;
+const EXPECTED_UNKOWN_SIZE: usize = 0x378; // Used only for optimization.
+
+fn partition_rbf(
+    id: PathBuf,
+    rbf: &[u8],
+) -> Result<Vec<SectionInfo>, (Vec<SectionInfo>, &'static str)> {
```

Comment:
If there's an error, the current progress on parsing it is not all that useful to us; I think it'd be simpler to return `Result<Vec<SectionInfo>, &'static str>`

Reply:
I disagree, the current progress is very useful in identifying where it failed. Currently the tool dumps the progress into the `section_infos` file. The offset of the final `Unknown` allows one to know which part of the file they need to examine (something which was very useful when debugging.)

Reply:
If there is an error here, it means the bitstream is in some way corrupt. At that point you might as well `panic` with an error to the human on the other end instead of returning an incomplete array and a string that Rust code can't read.